src/Hadolint/Rules.hs

```diff
@@ -1284,3 +1292,178 @@ useraddFlagL = instructionRule code severity message check
     hasLFlag = Shell.hasAnyFlag ["l", "no-log-init"]
     hasUFlag = Shell.hasAnyFlag ["u", "uid"]
     hasLongUID cmd = any ((> 5) . Text.length) (Shell.getFlagArg "u" cmd)
+
+noInvalidLabelKey :: Rule
+noInvalidLabelKey = instructionRule code severity message check
+  where
+    code = "DL3047"
+    severity = DLStyleC
+    message = "Invalid label key."
+    check (Label pairs) = hasNoInvalidKey pairs
+    check _ = True
+
+    hasNoInvalidKey prs = null $ [(l, v) | (l, v) <- prs,
+                                          Text.head l `notElem` ['a'..'z'] ||
+                                          Text.last l `notElem` ['a'..'z'] ++ ['0'..'9'] ||
+                                          containsIllegalChar l ||
+                                          hasReservedNamespace l ||
+                                          hasConsecutiveSeparators l]
+    containsIllegalChar :: Text.Text -> Bool
+    containsIllegalChar l = any (`notElem` validChars) (Text.unpack l)
+    hasReservedNamespace :: Text.Text -> Bool
+    hasReservedNamespace l = "com.docker." `Text.isPrefixOf` l
+      || "io.docker." `Text.isPrefixOf` l
+      || "org.dockerproject." `Text.isPrefixOf` l
+    hasConsecutiveSeparators :: Text.Text -> Bool
+    hasConsecutiveSeparators l = ".." `Text.isInfixOf` l || "--" `Text.isInfixOf` l
+    validChars = ['0'..'9'] ++ ['a' .. 'z'] ++ ['.', '-']
+
+labelRuleGenerator :: Map.Map Text.Text Text.Text -> Bool -> [Rule]
+labelRuleGenerator labelschema strictlabels = concat
+  [
+    fmap missingLabelRule (Map.keys labelschema),
+    fmap labelIsNotEmptyRule (Map.keys labelschema),
+    fmap labelIsNotUrlRule (Map.keys (Map.filter (\v -> v == Text.pack "url") labelschema)),
```

Comment:
There is no need for `Text.pack` here, afaik. That's because of `OverloadedStrings`, which will convert automatically any string literal into the right type for you

Reply:
Thats right. This nicely reduces to `Map.filter (== "url")`